Help message in `E0369` points to wrong file

<!--
Thank you for filing a bug report! 🐛 Please provide a short summary of the bug,
along with any information you feel relevant to replicating the bug.
-->

I have the following code, split into two files:

```rust
// src/main.rs
// this appears in the error

mod name;

use name::Name;

fn main() {
    let name_1 = Name::from("John Cena");
    let name_2 = Name::from("Not John Cena");
    println!("{}", name_1 == name_2);
}
```

```rust
// src/name.rs
pub struct Name {
    pub first: String,
    pub middle: String,
    pub last: String,
}

impl From<&str> for Name {
    fn from(name: &str) -> Self {
        let mut parts = name.split_whitespace();
        let first = parts.next().unwrap_or_default();
        let last = parts.clone().last().unwrap_or_default();
        let middle = parts.collect::<Vec<_>>().join(" ");
        Self {
            first: first.to_owned(),
            middle,
            last: last.to_owned(),
        }
    }
}
```

This code, as expected, produces `E0369`, shown below. However, the help message at the bottom of the error appears to be referencing line 2 in `src/main.rs`, while displaying the filename `src/name.rs`, thus suggesting we add the `derive` annotation to a comment in `src/main.rs`. It should instead point to the `struct` in `src/name.rs`.

```rs
error[E0369]: binary operation `==` cannot be applied to type `Name`
  --> src/main.rs:11:27
   |
11 |     println!("{}", name_1 == name_2);
   |                    ------ ^^ ------ Name
   |                    |
   |                    Name
   |
note: an implementation of `PartialEq` might be missing for `Name`
  --> src/name.rs:2:1
   |
2  | pub struct Name {
   | ^^^^^^^^^^^^^^^ must implement `PartialEq`
help: consider annotating `Name` with `#[derive(PartialEq)]`
  --> src/name.rs:2:1
   |
2  + #[derive(PartialEq)]
3  | // this appears in the error
   |
```

### Meta
<!--
If you're using the stable version of the compiler, you should also check if the
bug also exists in the beta or nightly versions.
-->

`rustc --version --verbose`:
```
rustc 1.73.0-nightly (903e279f4 2023-07-18)
binary: rustc
commit-hash: 903e279f468590fa3425f8aff7f3d61a5a873dbb
commit-date: 2023-07-18
host: x86_64-unknown-linux-gnu
release: 1.73.0-nightly
LLVM version: 16.0.5
```

<!-- TRIAGEBOT_START -->

<!-- TRIAGEBOT_ASSIGN_START -->

<!-- TRIAGEBOT_ASSIGN_DATA_START$${"user":"clubby789"}$$TRIAGEBOT_ASSIGN_DATA_END -->

<!-- TRIAGEBOT_ASSIGN_END -->
<!-- TRIAGEBOT_END -->

--- a/compiler/rustc_errors/src/emitter.rs
+++ b/compiler/rustc_errors/src/emitter.rs
@@ -1982,7 +1982,7 @@ impl EmitterWriter {
                 // We special case `#[derive(_)]\n` and other attribute suggestions, because those
                 // are the ones where context is most useful.
                 let file_lines = sm
-                    .span_to_lines(span.primary_span().unwrap().shrink_to_hi())
+                    .span_to_lines(parts[0].span.shrink_to_hi())
                     .expect("span_to_lines failed when emitting suggestion");
                 let line_num = sm.lookup_char_pos(parts[0].span.lo()).line;
                 if let Some(line) = file_lines.file.get_line(line_num - 1) {

--- a/tests/ui/modules/issue-107649.stderr
+++ b/tests/ui/modules/issue-107649.stderr
@@ -11,7 +11,7 @@ help: consider annotating `Dummy` with `#[derive(Debug)]`
    --> $DIR/auxiliary/dummy_lib.rs:2:1
     |
 2   + #[derive(Debug)]
-3   | #[path = "auxiliary/dummy_lib.rs"]
+3   | pub struct Dummy;
     |
 
 error: aborting due to previous error

